velox/functions/lib/aggregates/noisy_aggregation/sketch/SfmSketch.cpp

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/lib/aggregates/noisy_aggregation/sketch/SfmSketch.h"
+#include <cmath>
+#include "velox/common/base/BitUtil.h"
+#include "velox/common/base/Exceptions.h"
+#include "velox/common/memory/HashStringAllocator.h"
+
+namespace facebook::velox::functions::aggregate {
+
+using Allocator = facebook::velox::StlAllocator<int8_t>;
+
+namespace {
+// Internal validation functions - not exposed in header
+void validateEpsilon(double epsilon) {
+  VELOX_CHECK(
+      epsilon > 0,
+      "Epsilon must be greater than zero or equal to kNonPrivateEpsilon");
+}
+
+void validatePrecision(uint32_t precision, uint32_t indexBitLength) {
+  VELOX_CHECK(
+      precision + indexBitLength <= 64,
+      "Precision + indexBitLength cannot exceed 64");
+}
+
+void validatePrefixLength(uint32_t indexBitLength) {
+  VELOX_CHECK(
+      indexBitLength >= 1 && indexBitLength <= 16,
```

Comment:
Split this check into 2 for better error messages.

Use VELOX_CHECK_EQ, _GT, _LE, etc. here and in other places.